dev/generate-cuda.py

```diff
@@ -81,7 +81,10 @@ def getthread_dim(pos):
     return code
 
 
-def traverse(node, args={}, forvars=[], declared=[]):
+def traverse(node, args=None, forvars=None, declared=None):
+    args = args or {}
+    forvars = forvars or []
+    declared = declared or []
```

Comment:
Was the code depending on the fact these were mutable defaults?

Reply:
First, I'll ping the author, @reikdas, but I'll also look into it.

Reply:
I've attempted to make all passing explicit to see if that fixes it. Since the first call uses the defaults, all modifications after that change the default globally, so it's like passing these in when they are not listed there. If I get it working, @reikdas should still look it over.

Reply:
Of the three, `forvars` and `declared` are mutated in place, which would have been a problem if `traverse` were called multiple times without somehow clearing those arguments. (I'm not really sure how you'd get access to the objects; maybe in `__code__` or something.)

The `args` variable could remain a dict argument, not that it matters all that much. But it's not safe for `forvars` and `declared` to be list arguments.

Good catch, whichever automated test this was. It's a highly recursive function—it could be difficult to walk to ensure that a caller never changes something in place.

Reply:
I have a personal pet peeve against Python's "truthiness," which is used in

```python
forvars = forvars or []
```

I would say

```python
if forvars is None:
    forvars = []
```

You'll notice that I avoid putting an object in the predicate of an `if` statement unless it has `bool` type (as much as I'm aware, since I track types manually). In these expressions, the `or`-based assignment is clear enough, but in general it would be nice to always require explicit `· is not None`, `· != 0`, `len(·) != 0` checks.

Reply:
You get access by leaving the arguments out. Which is exactly what this did several times, for forvars and declared.

It's still bad for a function to modify it's mutable argument unless that is unmistakably part of the interface, but it's really bad if there's also a default. As a hard-and-fast rule: no mutable default arguments, ever (which is the contents of the check). 

```python
def f(x = []):
    x.append(1)
    print(x)

f() # -> [1]
f() # -> [1, 1]
```
This is what was happening when this was leaving off the final two arguments (which happened several times).

Reply:
> I've attempted to make all passing explicit to see if that fixes it. Since the first call uses the defaults, all modifications after that change the default globally, so it's like passing these in when they are not listed there. If I get it working, @reikdas should still look it over.

Our messages crossed. In general, it's not a bad technique for the entry of a recursive call to create mutable lists and for these lists to be mutably appended through it, but your last commit was necessary to make that work, now that they're not global.

Functionally passing immutable objects down a recursive walk through a tree makes the changes apply to subtrees; non-functionally passing a mutable object and modifying it in place through the recurisve walk makes changes apply in the "reading order" through the tree. This is a walk over an AST of some code, where both cases can be relevant.

For instance, if you were walking over an AST of this code:

```python
def f(x):
    y = (x + 1)**2
    return y + 1
```

passing an immutable thing into the first Assign node, it wouldn't exist in the Return, but if you are in-place updating a list or dict, it's still available in the Return node. The immutable-passing is good for nested scopes, such as the scope of variables available in function `f`, but the mutable-passing is good for accumulating things, like which variables exist by the time you get to a particular node.

So it's not in principle bad that these variables are being changed in-place, if the latter feature is desired. Those variables do, however, have to be cleared or created anew at the beginning of recursion and always passed through all other steps of recursion. For things like this, I often separate the "entry point" of recursion with one function and the "continue walking" with another function. The "continue walking" function shouldn't have default arguments (to avoid errors like the one that Henry fixed), but the "entry point" often should, for a nice API.

Reply:
You have to be careful with Python truthiness, but there are specific places where it's safer than comparisons - for example, `==` is much more likely to throw than a truthiness check. I admit it's not as good as Ruby (where only False and Nil are "falsy"). In this case, the `or` trick kept me from having to repeat myself, though `forvars = [] if forvars is None else []` would make this completely explicit without providing too much extra boilerplate that a reader has to slog through. I'd be happy with either one. The `or` trick is seen in Python occasionally and is technically safe for any Optional collection, but it's more common in Ruby (due to the better true/false handling).

There's a PEP to add None aware operators, not sure if it ever will make it, but that would be nice.

Reply:
> As a hard-and-fast rule: no mutable default arguments, ever (which is the contents of the check).

That is a good hard-and-fast rule. (Our messages crossed again.) You'd get that if the "entry point" is one function, with defaults, and the "continue walking" is another function. That "entry point" function probably doesn't even need `forvars` and `declared`, since they're probably just temporary variables for the recursion, not something that the interface needs.

This two-functions-for-recursion pattern comes up a lot. That was behind the motivation for separating the `ak.Array` interface from the `Content` tree nodes (separating the data structure into an "entry point" and "recursively composable" parts).

We probably won't hear from @reikdas until morning in India.  `:)`

Reply:
> In general, it's not a bad technique for the entry of a recursive call to create mutable lists and for these lists to be mutably appended through it, but your last commit was necessary to make that work, now that they're not global.

No, I only mean you should never depend on Python's default arguments for this. It's perfectly fine to create and pass a mutable structure (that's what I mean by "unless that is unmistakably part of the interface"). If it's clear that it's going to modify something in place, then that's fine. But mutable argument _defaults_ are very unexpected, and will cause hidden and very strange effects for the caller.

Reply:
Setting None as the default, and letting "None" start the chain is fine too. Though two functions make the interface tighter and explicit, which is better. :)

Reply:
It seems like there is a rather big bug in the code. I originally intended to pass the default values to `traverse` when not explicitly specified, i.e. `{}, [], []` but that is clearly not how it is supposed to work in Python. But doing that also does not work, so there is a logical error in my code somewhere that was being hidden. 

Sorry for the trouble. I'll try and find a fix ASAP (perhaps I won't be able to provide it tonight, but I should be able to get a fix out by tomorrow). 

Thanks for finding this! And also thanks for the above discussion - it is really relevant for my current project as well!

Reply:
Hi @reikdas! It's not a giant rush, though this PR will have to wait for your fix.

It's funny that just after having this whole conversation, I ran into [the same issue on a StackOverflow answer](https://stackoverflow.com/a/59898173/1623645). I'll be a good citizen and edit that answer.

Reply:
I've put a `# noqa: B006` here so I can verify that this passes tests in its original form, and it's not something else causing problems (which is unlikely). If it passes, we could probably either get this in and then make a PR that removes the `noqa` comment and properly fixes this in the near future. Might need a `# TODO` or something like that to be extra sure it does not get missed? :)

It's a common mistake; it seems like it should work unless you realize that the function definition only happens once, including default arguments. And, if you don't mutate the arguments, it doesn't harm you - but as soon as you do, it's becomes very confusing.

Reply:
This script is only executed when generating CUDA code, which is developer-initiated, not very frequent, and not permanent. The CUDA that it generated is going to be a "first draft" of the CUDA kernels as hand-written code.

So yes, just marking this `noqa` and moving on would also be acceptable, but we should remember this issue before the next time we generate CUDA code.